File: emlun-scala/day11.scala
object Day11 extends App {

  val moves: List[String] = (for {
    line <- io.Source.stdin.getLines()
    move <- line.trim.split(raw",")
  } yield move).toList

  def execute(state: (Int, Int), move: String): (Int, Int) = state match {
    case (x, y) => move match {
      case "n"  => (x, y + 2)
      case "ne" => (x + 1, y + 1)
      case "se" => (x + 1, y - 1)
      case "s"  => (x, y - 2)
      case "sw" => (x - 1, y -1)
      case "nw" => (x - 1, y + 1)
    }
  }

  def dist(pos: (Int, Int)): Int = pos match {
    case (x, y) => {
      val ax = Math.abs(x)
      val ay = Math.abs(y)
      val diagonal: Int = Math.min(ax, ay)

      diagonal + (ax - diagonal) + (ay - diagonal) / 2
    }
  }

  def solveA(moves: List[String]): Int = dist(moves.foldLeft((0, 0))(execute))
  def solveB(moves: List[String]): Int =
    moves
      .scanLeft((0, 0))(execute)
      .map(dist)
      .foldLeft(0)(Math.max)

  println(s"A: ${solveA(moves)}")
  println(s"B: ${solveB(moves)}")
}

File: erikdyrelius-python3/day11.py
from aocbase import readInput

inp = readInput()

def dist(x, y):
    if x > 0:
        y = max(y-x, 0)
    elif x < 0:
        y = min(y-x, 0)
    return abs(x)+abs(y)

d = {
    "n":(0, 1),
    "ne":(1, 1),
    "se":(1, 0),
    "s":(0, -1),
    "sw":(-1, -1),
    "nw":(-1, 0)
}

x, y = 0, 0
mx = 0

for s in inp.split(","):
    x += d[s][0]
    y += d[s][1]
    mx = max(mx, dist(x, y))
print("Puzzle 1 = {}, Puzzle 2 = {}".format(dist(x, y), mx))